src/Models/BaseElement.php

```diff
@@ -721,21 +721,38 @@ public function CMSEditLink($directLink = false)
             return $link;
         }
 
-        if (!$page instanceof SiteTree && method_exists($page, 'CMSEditLink')) {
-            $link = Controller::join_links($page->CMSEditLink(), 'ItemEditForm');
+        if (!$page instanceof SiteTree) {
+            if (ClassInfo::hasMethod($page, 'CMSEditLink')) {
+                $link = Controller::join_links($page->CMSEditLink(), 'ItemEditForm');
+            } else {
+                $link = null;
+                $this->extend('updateCMSEditLink', $link);
```

Comment:
Shouldn't this extension hook still be be called even if `ClassInfo::hasMethod($page, 'CMSEditLink')` ?

Reply:
Probably, not yet. Here we just generate only a some part of the link. Further we use this part to generate the rest part of the link and after that we use $this->extend('updateCMSEditLink', $link);

Reply:
seems like it should be 
```
$link = null;
if (ClassInfo::hasMethod($page, 'CMSEditLink')) {
    $link = Controller::join_links($page->CMSEditLink(), 'ItemEditForm');
}
$this->extend('updateCMSEditLink', $link);
```
Any custom code implementing the link can add a `if (is_null($link))` check and proceed from there.  It's entirely possible if $link is non-null they still need to update the link to something else

Reply:
actually it's a bit confusing that $link can be set multiple times in this method, for instance

```php
        if (!($page instanceof SiteTree)) {
            if (ClassInfo::hasMethod($page, 'CMSEditLink')) {
                $link = Controller::join_links($page->CMSEditLink(), 'ItemEditForm');
            } else {
                $link = null;
                $this->extend('updateCMSEditLink', $link);
                return $link;
            }
        } else {
            $link = $page->CMSEditLink();
        }

        // In-line editable blocks should just take you to the page. Editable ones should add the suffix for detail form
        if (!$this->inlineEditable() || $directLink) {
            if (!($page instanceof SiteTree)) {
                $link = Controller::join_links(
                    $link,
                    'field',
                    $relationName,
                    'item',
                    $this->ID,
                    'edit'
                );
```

ideally we'd have something like 

```php
$link = null
if ($conditionA) {
  $link = ...
} elseif ($conditionB) {
  $link = ...
}
$this->extend('updateCMSEditLink', $link);
return $link;
```